pkg/search/postgres/store_cache.go

```diff
@@ -346,6 +346,33 @@ func (c *cachedStore[T, PT]) Walk(ctx context.Context, fn func(obj PT) error) er
 	return c.walkCacheNoLock(ctx, fn)
 }
 
+// GetByQueryFn iterates over the objects from the store matching the query.
+func (c *cachedStore[T, PT]) GetByQueryFn(ctx context.Context, query *v1.Query, fn func(obj PT) error) error {
+	defer c.setCacheOperationDurationTime(time.Now(), ops.GetByQuery)
+	identifiers, err := c.underlyingStore.GetIDsByQuery(ctx, query)
+	// Fallback to the underlying store on error.
+	if err != nil {
+		log.Errorf("Failed to get identifiers by query, falling back to get results by query: %v", err)
+		return c.underlyingStore.GetByQueryFn(ctx, query, fn)
+	}
+	c.cacheLock.RLock()
+	defer c.cacheLock.RUnlock()
+	for _, id := range identifiers {
+		obj, found := c.cache[id]
+		if !found {
+			log.Warnf("Object %q not found in store cache", id)
+			continue
+		}
+		if !c.isReadAllowed(ctx, obj) {
+			continue
+		}
+		if err := fn(obj); err != nil {
+			return err
+		}
+	}
```

Comment:
Could some or all of this be moved to a helper since we repeat it in other places?

Reply:
Done